### PR TITLE
Adding ability to pass constraint function to matcher

### PR DESCRIPTION
**Pull request by joshuacullenlux**
This will let you pass a function to any matcher. The function will be run with the matchers path and values. 

This will enable us to add custom constraints. 

For example we could have optional fields that are required if one of them exists.

```
// AND
var values = {
   street_number: 1
}

var func =  function(path, val) {
  if(val.street_number && !val.street_name) {
    return [{
      error: "If street_number is defined we require the street name"
    }]
}

const errors = m.match('path', values, func)

```

### CODE REVIEW

File: lib/matchers/objectWithOnly.js [hunk 1]
```diff
@@ -28,6 +33,17 @@ module.exports = factory({
           });
         }
       }
+
+      if (this.constraints) {
+        var constraintErrors = this.constraints(path, val)
+
+        if (constraintErrors) {
+          errors = errors.concat(
```
Comment (nguyenchr) [inline lib/matchers/objectWithOnly.js]: should just concat `constraintErrors`

File: lib/constraints.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+var Constraint = Object.create({
```
Comment (nguyenchr) [inline lib/constraints.js]: I don't think this gets used

File: README.md [hunk 3]
```diff
@@ -232,6 +233,41 @@ s.assert(person, {
...
+Currently only objectWithOnly is supported.
+
+An example use case is related optional fields
+
+```js
+// AND relationship between two optinal fields
+
+var constraintFunc = function (path, value) {
+  if (value.street_number && !value.post_code) {
+    return {
```
Comment (nguyenchr) [inline README.md]: should return an array of errors

Comment (nguyenchr) [reply README.md]: and return empty array if there are no errors

File: README.md [hunk 4]
```diff
@@ -232,6 +233,41 @@ s.assert(person, {
...
+## Custom constraints
+
+Custom constraints can be applied by passing a function as the second argument when creating the schema. This function will be run on match and you are able to return an array of errors.
+
+Currently only objectWithOnly is supported.
+
+An example use case is related optional fields
+
+```js
+// AND relationship between two optinal fields
```
Comment (nguyenchr) [inline README.md]: s/optinal/optional

File: README.md [hunk 5]
```diff
@@ -232,6 +233,43 @@ s.assert(person, {
...
+  street_number: new number({optional: true}),
+  post_code: new number({optional: true}),
+}, constraintFunc);
+
+var value = {
+  email_address: 'test@strummer.com',
+  street_number: 12,
+}
+
+const errors = schema.match(vals)
```
Comment (deancouch) [inline README.md]: s/vals/value

Comment (deancouch) [reply README.md]: looks good, just a minor typo here

**Pull request merged**